core/data_category.js

```diff
@@ -0,0 +1,429 @@
+/**
+ * @license
+ * Visual Blocks Editor
+ *
+ * Copyright 2017 Google Inc.
+ * https://developers.google.com/blockly/
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Data Flyout components including variable and list blocks.
+ * @author marisaleung@google.com (Marisa Leung)
+ */
+'use strict';
+
+/**
+ * @name Blockly.DataCategory
+ * @namespace
+ **/
+goog.provide('Blockly.DataCategory');
+
+goog.require('Blockly.Blocks');
+goog.require('Blockly.VariableModel');
+goog.require('Blockly.Variables');
+goog.require('Blockly.Workspace');
+
+/**
+ * Construct the blocks required by the flyout for the variable category.
+ * @param {!Blockly.Workspace} workspace The workspace containing variables.
+ * @return {!Array.<!Element>} Array of XML block elements.
+ */
+Blockly.DataCategory = function(workspace) {
+  var variableModelList = workspace.getVariablesOfType('');
+  variableModelList.sort(Blockly.VariableModel.compareByName);
+  var xmlList = [];
+
+  Blockly.DataCategory.addCreateButton(xmlList, workspace, 'VARIABLE');
+
+  for (var i = 0; i < variableModelList.length; i++) {
+    Blockly.DataCategory.addDataVariable(xmlList, variableModelList[i]);
+  }
+
+  if (xmlList.length > 1) { // The button is always there.
```

Comment:
Check variableModelList.length here, the way you do below, and update the comments to say // Only show these blocks if there is at least one variable [list] defined.